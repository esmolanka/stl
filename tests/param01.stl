type Maybe a =
  < Nothing
  , Just : a
  >
#eval Maybe
#eval Maybe Unit


type Either a b =
  < Left  : a
  , Right : b
  >
#eval Either
#eval Either Unit
#eval Either Unit (Maybe Unit)


type These a b =
  < This  : a
  , That  : b
  , These : {this : a, that : b}
  >
#eval These
#eval These Unit
#eval These Unit (Maybe Unit)
