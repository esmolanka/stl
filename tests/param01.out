tests/param01.types:5:1-5:12:
  Normalised:
    λ a. ∀ (ρ : Row). <Nothing? : Unit | Just? : a | ... ρ>
  Kind:
    Type -> Type
tests/param01.types:6:1-6:17:
  Normalised:
    ∀ (ρ : Row). <Nothing? : Unit | Just? : Unit | ... ρ>
  Kind:
    Type
tests/param01.types:13:1-13:13:
  Normalised:
    λ a. λ b. ∀ (ρ : Row). <Left? : a | Right? : b | ... ρ>
  Kind:
    Type -> Type -> Type
tests/param01.types:14:1-14:18:
  Normalised:
    λ b. ∀ (ρ : Row). <Left? : Unit | Right? : b | ... ρ>
  Kind:
    Type -> Type
tests/param01.types:15:1-15:30:
  Normalised:
    ∀ (ρ : Row).
      < Left? : Unit
      | Right? : ∀ (ρ : Row). <Nothing? : Unit | Just? : Unit | ... ρ>
      | ... ρ
      >
  Kind:
    Type
tests/param01.types:23:1-23:12:
  Normalised:
    λ a. λ b. ∀ (ρ : Row).
      < This? : a
      | That? : b
      | These? : ∃ (ρ : Row). {this : a, that : b | ρ}
      | ... ρ
      >
  Kind:
    Type -> Type -> Type
tests/param01.types:24:1-24:17:
  Normalised:
    λ b. ∀ (ρ : Row).
      < This? : Unit
      | That? : b
      | These? : ∃ (ρ : Row). {this : Unit, that : b | ρ}
      | ... ρ
      >
  Kind:
    Type -> Type
tests/param01.types:25:1-25:29:
  Normalised:
    ∀ (ρ : Row).
      < This? : Unit
      | That? : ∀ (ρ : Row). <Nothing? : Unit | Just? : Unit | ... ρ>
      | These? :
          ∃ (ρ : Row).
            { this : Unit
            , that : ∀ (ρ : Row). <Nothing? : Unit | Just? : Unit | ... ρ>
            | ρ
            }
      | ... ρ
      >
  Kind:
    Type